dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
```

Comment:
Just added fucntional.h  and: cstd/functional.h:24:56: error: backslash-newline at end of file [-Werror]
 #define DECLARE_PREDICATE(name, state_type, item_type) \
                                                        ^
cc1: all warnings being treated as errors